### PR TITLE
add task: Developing and debugging services with local IDE inside container

### PR DESCRIPTION
**Pull request by lyzhang1999**
There is only one introductory article on how to develop and debug services on Kubernetes(telepresence : https://kubernetes.io/docs/tasks/debug-application-cluster/local-debugging/). I want to introduce a new way to develop in the Kubernetes environment: Nocalhost. The project has entered CNCF Landscape . There are more than 600 stars on Github. Hope to help more people simplify the development work in the Kubernetes environment.

I think how to develop better on Kubernetes is about the core of Kubernetes, Just like telepresence introductory, Nocalhost is now another way for development, I hope the community can accept it.

### CODE REVIEW

**Review by sftim — COMMENTED** [review]
This looks like it promotes a specific 3rd party tool: [Nocalhost](https://nocalhost.dev/)

That kind of mention / promotion is [not allowed](https://kubernetes.io/docs/contribute/style/content-guide/#what-s-allowed) here. However, it'd be OK to publish a similar page within https://nocalhost.dev/

/hold

**Comment by lyzhang1999** [comment]
> This looks like it promotes a specific 3rd party tool: [Nocalhost](https://nocalhost.dev/)
> 
> That kind of mention / promotion is [not allowed](https://kubernetes.io/docs/contribute/style/content-guide/#what-s-allowed) here. However, it'd be OK to publish a similar page within https://nocalhost.dev/
> 
> /hold

but how about telepresence https://kubernetes.io/docs/tasks/debug-application-cluster/local-debugging/

**Comment by sftim** [comment]
> how about telepresence

We could replace https://kubernetes.io/docs/tasks/debug-application-cluster/local-debugging/ with a page that lists 3rd party tools for local development workflows and outlines a task to carry out using any compatible tool.

**Comment by lyzhang1999** [comment]
> > how about telepresence
> 
> We could replace https://kubernetes.io/docs/tasks/debug-application-cluster/local-debugging/ with a page that lists 3rd party tools for local development workflows and outlines a task to carry out using any compatible tool.

can't agree anymore

**Comment by tengqm** [comment]
@lyzhang1999 PLease separate the changes to English from changes to Chinese.
Always commit to English upstream first, please.

**Comment by lyzhang1999** [comment]
> @lyzhang1999 PLease separate the changes to English from changes to Chinese.
> Always commit to English upstream first, please.

Done, Thank you!

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (tengqm) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:2]: 这个表达很有问题。
是 IDE 在容器中么？

Comment (tengqm) [reply content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:2]: 后文的英文表达问题太多，请寻求英文较好的同事帮助重写。

Comment (lyzhang1999) [reply content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:2]: 好的，我重写一下

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:2]:
```suggestion
title: Developing and debugging applications running in your cluster
```

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:7]:
```suggestion

This page describes how you can external tools to debug an application that's running in your cluster.
The tools covered are:
{{% thirdparty-content %}}

- [Nocalhost](https://nocalhost.dev)

```

It would be helpful to make a similar change to the Telepresence task, so that they match each other.

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:16]:
```suggestion
* {{< include "task-tutorial-prereqs.md" >}}
```

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:17]:
```suggestion
* A local IDE: either Microsoft Visual Studio Code, or Jetbrains
* Nocalhost itself. To learn how to install Nocalhost, visit [Installation](https://nocalhost.dev/eng/installation/)
```

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:21]: I wouldn't be confident I could explain these steps to someone else. What should a reader do?

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:29]: This sounds like a task prequisite (you don't need it to perform debugging, but the task doesn't make sense without it). I'd explain this earlier.

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:33]: “Wait bookinfo becomes ready” isn't correct English; can you rewrite it?

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:39]:
```suggestion
## Modifiying your application
```

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:52]: Is VS Code a strict prerequisite? It's not clear?

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:91]: From the Kubernetes project's point of view, we'd want to signpost readers to:
```suggestion
* Learn about [Developing and debugging services locally](/docs/tasks/debug-application-cluster/local-debugging/)
```

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:83]: I'd omit this text. It's not helping the reader learn a task.

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:75]:
```suggestion
In your web browser, refresh the window or tab where you have `http://127.0.0.1:39080/` open. You will see the following interface:
```

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:61]: [Don't include the prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt)

**Review by sftim — CHANGES_REQUESTED** [review]
Thanks for your work so far on this PR @lyzhang1999
With extra rewriting we might be able to accept this. It will need more changes than I've indicated so far.

Please refer to the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/) to understand how we write documentation for Kubernetes.

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (lyzhang1999) [reply content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:61]: done

**Comment by lyzhang1999** [comment]
> Thanks for your work so far on this PR @lyzhang1999
> 
> With extra rewriting we might be able to accept this. It will need more changes than I've indicated so far.
> 
> 
> 
> Please refer to the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/) to understand how we write documentation for Kubernetes.

Done, please help me to review again, thank you

**Comment by lyzhang1999** [comment]
Nocalhost Sandbox proposal: https://github.com/cncf/tag-app-delivery/issues/142

**Comment by lyzhang1999** [comment]
@sftim @tengqm Please review this PR, Thanks.

File: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md
Comment (tengqm) [inline content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:15]: I still believe this post needs some work. For example, the paragraph above could be revised:

```suggestion
Nocalhost is a tool that enables developers to write applications to run on a Kubernetes Cluster.
By using a Nocalhost IDE (for example, VSCode or Jetbrains), you can switch the remote workload
into development mode. When code is updated on your local machine, changes to code are automatically
synchronized to the remote development environment. The code change then automatically triggers
an update to the target application. Depending on the nature of the application, such an update could
be a hot reload or a process restart. The development container where the updated application runs
inherits all the configurations (e.g. ConfigMap, Secret, Volume, Env) from the original container.
```

I'm not sure what a "development container" means and I'm curious how the target workload/application
is automatically restarted.

Comment (tengqm) [reply content/en/docs/tasks/debug-application-cluster/inside-container-developing.md:15]: I'm not a native speaker, so the rewriting above only serves an example how the text can be reorganized. It is only an example to show how heavily this post should be revised. Without such a rewriting, this post is difficult to understand.

Comment (lyzhang1999) [reply content/en/docs/tasks/debug-application-cluster/inside-container-developing.md]: Thank you! 
I have explained about the "development container" and rewritten it according to your comments.

**Comment by lyzhang1999** [comment]
@tengqm Done

**Comment by lyzhang1999** [comment]
@sftim Done

**Comment by sftim** [comment]
The [review](https://github.com/kubernetes/website/pull/28564#pullrequestreview-689419546) I wrote in June still applies 
@lyzhang1999. 

I don't see any grounds for reconsidering that viewpoint. See https://github.com/kubernetes/enhancements/tree/master/keps/sig-docs/1326-third-party-content-in-docs for some more context.

**Comment by sftim** [comment]
I propose closing this PR: the change isn't acceptable as it's not in line with Kubernetes policies on third party content.

**Comment by lyzhang1999** [comment]
> The [review](https://github.com/kubernetes/website/pull/28564#pullrequestreview-689419546) I wrote in June still applies @lyzhang1999.
> 
> I don't see any grounds for reconsidering that viewpoint. See https://github.com/kubernetes/enhancements/tree/master/keps/sig-docs/1326-third-party-content-in-docs for some more context.

Nocalhost has now been donated to CNCF and he is not affiliated with any vendor.
Secondly, development in a K8s environment is a common problem for almost all companies using K8s. Developing in a K8s environment is a pain and development is inefficient.
I hope the community will be receptive to suggestions for other solutions besides Telepresence.

**Comment by lyzhang1999** [comment]
Just like collecting K8s logs, developers also need to understand how to work efficiently and cheaply in a K8s environment.
Nocalhost is now becoming the new choice for development in K8s, and as the maintainer of the documentation, we should put ourselves in the shoes of the end user.

**Comment by sftim** [comment]
> We could replace https://kubernetes.io/docs/tasks/debug-application-cluster/local-debugging/ with a page that lists 3rd party tools for local development workflows and outlines a task to carry out using any compatible tool.

&mdash; https://github.com/kubernetes/website/pull/28564#issuecomment-866005623

How about doing that? I'm happy to explain why I recommend this.

**Comment by lyzhang1999** [comment]
> > We could replace https://kubernetes.io/docs/tasks/debug-application-cluster/local-debugging/ with a page that lists 3rd party tools for local development workflows and outlines a task to carry out using any compatible tool.
> 
> — [#28564 (comment)](https://github.com/kubernetes/website/pull/28564#issuecomment-866005623)
> 
> How about doing that? I'm happy to explain why I recommend this.

I think it's a good idea for end users to at least find different development options in the K8s environment,  Thank you for thinking of the end user.

**Pull request closed without merging**
